### PR TITLE
added proteingym data to evedesign system conversion fxn

### PR DESCRIPTION
**Pull request by ethane4**
Left a bunch of comments in for discussion.

Added the proteingym-base repo as an evedesign dependency + a src/proteingym.py script that handles translation from loaded pgdata Subset to evedesign system and labelledinstancedataset.

I also tried to keep this as light/versatile as possible to offload as much work onto individual model entrypoints as possible ex. specifying split names/handling supervised vs unsupervised/etc.

### CODE REVIEW

File: src/evedesign/proteingym.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+from evedesign.structure import Structure
+from evedesign.system import EntityInstance, Protein, System, SystemInstance
+
+from proteingym.base import Dataset, Subsets
+from proteingym.base.sequence import SequenceType
+
+
+def wildtype_or_none(dataset: Dataset) -> str | None:
+    """
+    Extract wildtype sequence from dataset if available to set Entity rep, which
```
Comment (thomashopf) [inline src/evedesign/proteingym.py]: please update all docstrings in this file to full numpy format

File: src/evedesign/proteingym.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    if len(dataset.msas) == 0:
+        return None
+
+    # this case is non-trivial scientifically yet highly relevant; discuss first before implementing...
+    if dataset.msas[0].sequence_start is not None and dataset.msas[0].sequence_start != 1:
+        raise NotImplementedError(
+            "MSA region start different from 1"
+        )
+
+    # take first MSA by default for now
```
Comment (thomashopf) [inline src/evedesign/proteingym.py:60]: we'll need some extra checking here that the alignment has in fact a3m format

File: src/evedesign/proteingym.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+
+def update_structure(atom_array: AtomArray) -> Structure:
+    """
+    Wrap a ProteinGym structure (a biotite ``AtomArray``) into the evedesign
+    ``Structure`` model -- assumes monomers
+    """
+    s = Structure(atom_array)
+    assert len(s.chains()) == 1
+    return s.get_chain(s.chains()[0])
```
Comment (thomashopf) [inline src/evedesign/proteingym.py]: should be able to return `s` directly as-is without needing to call get_chain

File: src/evedesign/proteingym.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    # this case is non-trivial scientifically yet highly relevant; discuss first before implementing...
+    if dataset.msas[0].sequence_start is not None and dataset.msas[0].sequence_start != 1:
+        raise NotImplementedError(
+            "MSA region start different from 1"
+        )
+
+    # take first MSA by default for now
+    first_msa = dataset.msas[0].value
+
+    # TODO: length of weights and sequences not equal, presumaby invalid sequence
```
Comment (thomashopf) [inline src/evedesign/proteingym.py]: need to decide if we bring in weight handling - is this fixed in pg-base?

File: src/evedesign/proteingym.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+from proteingym.base import Dataset, Subsets
+from proteingym.base.sequence import SequenceType
+
+
+def wildtype_or_none(dataset: Dataset) -> str | None:
+    """
+    Extract wildtype sequence from dataset if available to set Entity rep, which
+    needs to be set to None otherwise
+
+    Note the reference_sequence computed field on Dataset fails with StopIteration error
```
Comment (thomashopf) [inline src/evedesign/proteingym.py]: worth double checking if this is fixed, or remove this comment

File: src/evedesign/proteingym.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    Build an evedesign System from a ProteinGym Dataset.
+
+    We default to a single-component protein system as there are no other cases
+    in ProteinGym (yet...) and assume structure numbering matches the rep
+    sequence and first_index by convention.
+    """
+    return System([
+        Protein(
+
+            first_index=1, # this is the case for 90% of the database, would
```
Comment (thomashopf) [inline src/evedesign/proteingym.py:139]: has this field been added to pg-base already?

File: src/evedesign/proteingym.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            },
+        )
+    ])
+
+
+def _labeled_dataset_from_df(df: pl.DataFrame, target: str) -> LabeledInstanceDataset:
+    """
+    Build a LabeledInstanceDataset from a ProteinGym dataframe slice.
+
+    TODO: need to handle insertion/deletion datasets properly; sequences are
```
Comment (thomashopf) [inline src/evedesign/proteingym.py]: let's bake this in as an explicit assertion until solved - raise error if instance rep length != system rep length

File: src/evedesign/proteingym.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+            # evedesign Structure model
+            structures={
+                struc.name: update_structure(struc.value)
+                for struc in dataset.structures
+            },
+        )
+    ])
+
+
+def _labeled_dataset_from_df(df: pl.DataFrame, target: str) -> LabeledInstanceDataset:
```
Comment (thomashopf) [inline src/evedesign/proteingym.py]: no need to prefix this with an underscore, we can expose this as a proper function I think

File: src/evedesign/proteingym.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+    subsets: Subsets,
+    split: str,
+    target: str,
+    test_fold: int | None = None,
+) -> tuple[System, LabeledInstanceTrainTestDataset]:
+    """
+    Map a ProteinGym dataset (subset) to evedesign representations, following the
+    train/test conventions of the benchmark training entrypoint.
+
+    Params
```
Comment (thomashopf) [inline src/evedesign/proteingym.py]: not fully matching numpy convention in docstring - should have ---- underneath and "Returns" instead of "Gives"

File: src/evedesign/proteingym.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+
+    valid_targets = [t.name for t in dataset.assay_targets]
+    if target not in valid_targets:
+        raise ValueError(
+            f"Target '{target}' is not present in dataset assay targets, "
+            f"valid options are: {', '.join(valid_targets)}"
+        )
+
+    if test_fold is None:
+        training_set = _labeled_dataset_from_df(dataset.to_df(), target)
```
Comment (thomashopf) [inline src/evedesign/proteingym.py]: I don't think it is safe to assume that the dataset should be the training dataset in all cases - might just as well be test set...

In that case would be cleaner to return a LabeledInstanceDataset instead

Also don't think pg-base/benchmark allow test_fold = None?

Let's discuss

File: src/evedesign/proteingym.py (hunk 2 above)
Comment (ethane4) [reply src/evedesign/proteingym.py:60]: Currently the only supported MSA format in pg-base is A3M. Unfortunately msa.format is an attribute tied to the manifest section and not the actual msa class in pg-base. I can go add it as an attribute if we think it is strictly necessary to do some check (should be fine to just assume it is always a3m w/ the current pg-base setup).

File: src/evedesign/proteingym.py (hunk 3 above)
Comment (ethane4) [reply src/evedesign/proteingym.py]: like literally just Structure(atom_array) ?

File: src/evedesign/proteingym.py (hunk 4 above)
Comment (ethane4) [reply src/evedesign/proteingym.py]: There are validation checks when building the dataset that confirm len(msa) = len(weights) if that is what you mean

File: src/evedesign/proteingym.py (hunk 5 above)
Comment (ethane4) [reply src/evedesign/proteingym.py]: This seems to be fixed when I test this with an example dataset

File: src/evedesign/proteingym.py (hunk 6 above)
Comment (ethane4) [reply src/evedesign/proteingym.py:139]: No we were arguing about it a bit. Henning's preference was to limit the number of user-provided dataset attributes as much as possible (ex. no mutated region, no start index, ...) On some level, the pg-benchmark wrappers will largely operate on plain sequence w/ no mutation strings - so I think we can ignore for now...?

File: src/evedesign/proteingym.py (hunk 7 above)
Comment (ethane4) [reply src/evedesign/proteingym.py]: So the current solution is to just store sequences in the dataframe in a3m format, which works for like 95% of our datasets, but we have some that would be pretty hard to infer correct a3m representations for (or I guess it would just depend on the pairwise aligner settings)

We can consider this solved if LabeledInstanceDataset can handle a3m-formatted seqs. I could also add a match state check if we think that is a good idea, but iterating the dataset would slow down the construction process

File: src/evedesign/proteingym.py (hunk 8 above)
Comment (ethane4) [reply src/evedesign/proteingym.py]: fixed

File: src/evedesign/proteingym.py (hunk 9 above)
Comment (ethane4) [reply src/evedesign/proteingym.py]: fixed

File: src/evedesign/proteingym.py (hunk 10 above)
Comment (ethane4) [reply src/evedesign/proteingym.py]: Yeah test fold is just an integer corresponding to the fold the user wants to select for test. I will just remove this case

File: src/evedesign/proteingym.py (hunk 2 above)
Comment (ethane4) [reply src/evedesign/proteingym.py:60]: also commented out this sequence_start check bc we took that attribute out of pg-base

Comment (thomashopf) [reply src/evedesign/proteingym.py:60]: ok, if default is now a3m then don't need to check

Is the underlying pg-base MSA class now able to handle sequences of different length or should we be reading from raw file here?

File: src/evedesign/proteingym.py (hunk 4 above)
Comment (thomashopf) [reply src/evedesign/proteingym.py]: Yes was referring to the different length of weight list vs. number of sequences in MSA; the way it is solved with the assertion is good and can remove the TODO note here

File: src/evedesign/proteingym.py [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+    Raises
+    ------
+    NotImplementedError
+        If the MSA region start is different from 1.
+    """
+    if len(dataset.msas) == 0:
+        return None
+
+    # this case is non-trivial scientifically yet highly relevant; discuss first before implementing...
```
Comment (thomashopf) [inline src/evedesign/proteingym.py]: we can remove the comment here if these attributes have been removed

File: src/evedesign/proteingym.py (hunk 6 above)
Comment (thomashopf) [reply src/evedesign/proteingym.py:139]: If all pieces are consistently numbered from 1 (including 3D structures), then fine with me to default to first_index=1 here

File: src/evedesign/proteingym.py (hunk 7 above)
Comment (thomashopf) [reply src/evedesign/proteingym.py]: Yes, the a3m-formatted sequences are the evedesign convention so if sequences in PG are formatted that way now then we can consider this solved

File: src/evedesign/proteingym.py (hunk 1 above)
Comment (ethane4) [reply src/evedesign/proteingym.py]: this should be resolved now

File: src/evedesign/proteingym.py (hunk 2 above)
Comment (ethane4) [reply src/evedesign/proteingym.py:60]: I believe I wrote in a match state check for confirming a3m validity but pg-base can handle variable insertion lengths yes. Also checked the resulting system[0][0].sequences.seqs that comes out of this new fxn and all looks good

File: src/evedesign/proteingym.py (hunk 4 above)
Comment (ethane4) [reply src/evedesign/proteingym.py]: removed

File: src/evedesign/proteingym.py (hunk 6 above)
Comment (ethane4) [reply src/evedesign/proteingym.py:139]: sounds good we can re-evaluate down the road if we do decide to alter this convention in pg-base

File: src/evedesign/proteingym.py (hunk 7 above)
Comment (ethane4) [reply src/evedesign/proteingym.py]: noted

File: src/evedesign/proteingym.py (hunk 11 above)
Comment (ethane4) [reply src/evedesign/proteingym.py]: yes the plan is for all MSAs to be directly generated from the overall sequence associated w/ each dataset, so we should never run into this issue

**Comment by ethane4** [comment]
Newest commit resolves all (?) conversations above + removes pycache files
